File: claude-code-agents-python/src/models/plugin_spec.py
"""Plugin specification models following anthropics/knowledge-work-plugins patterns."""

from __future__ import annotations

import json
from typing import Any, Optional

from pydantic import BaseModel, Field

__all__ = ["PluginSpec", "SkillSpec", "AgentSpec", "ConnectorSpec"]


class SkillSpec(BaseModel):
    """Specification for a single skill within a plugin."""

    name: str = Field(..., description="Skill name, used as filename stem")
    description: str = Field(default="", description="Human-readable description")
    frontmatter: dict[str, Any] = Field(
        default_factory=dict,
        description="YAML frontmatter fields (tools, model, etc.)",
    )
    content: str = Field(default="", description="Markdown body of the skill file")
    references: list[str] = Field(
        default_factory=list, description="File or URL references the skill uses"
    )
    scripts: list[str] = Field(
        default_factory=list, description="Shell scripts the skill can invoke"
    )

    def full_name(self) -> str:
        """Return the SKILL.md filename."""
        return f"{self.name}.md"


class AgentSpec(BaseModel):
    """Specification for an agent within a plugin."""

    name: str = Field(..., description="Agent identifier")
    description: str = Field(default="", description="What the agent does")
    tools: list[str] = Field(default_factory=list, description="Tools available to the agent")
    model: str = Field(default="claude-sonnet-4-20250514", description="Model to use")
    system_prompt: str = Field(default="", description="System prompt for the agent")

    def full_name(self) -> str:
        """Return the agent .md filename."""
        return f"{self.name}.md"


class ConnectorSpec(BaseModel):
    """Specification for an MCP connector."""

    name: str = Field(..., description="Connector identifier")
    type: str = Field(
        default="stdio",
        description="Transport type: stdio, sse, streamable-http",
    )
    server_config: dict[str, Any] = Field(
        default_factory=dict,
        description="MCP server configuration (command, args, env, etc.)",
    )
    placeholder_category: str = Field(
        default="",
        description="Category for ~~ placeholder substitution (e.g., 'api_key')",
    )

    def has_placeholders(self) -> bool:
        """Check if server config contains ~~ placeholder values."""
        config_str = json.dumps(self.server_config)
        return "~~" in config_str


class PluginSpec(BaseModel):
    """Full specification for a Claude Code plugin."""

    name: str = Field(..., description="Plugin name")
    version: str = Field(default="0.1.0", description="Semantic version")
    description: str = Field(default="", description="Plugin description")
    author: str = Field(default="", description="Plugin author")
    skills: list[SkillSpec] = Field(default_factory=list, description="Skills provided")
    agents: list[AgentSpec] = Field(default_factory=list, description="Agents provided")
    connectors: list[ConnectorSpec] = Field(
        default_factory=list, description="MCP connectors"
    )
    hooks: dict[str, list[dict[str, Any]]] = Field(
        default_factory=dict,
        description="Hook definitions keyed by event name",
    )
    lsp_servers: list[str] = Field(
        default_factory=list,
        description="LSP server binaries to configure",
    )

    @property
    def plugin_dir_name(self) -> str:
        """Directory name for the plugin."""
        return self.name.replace(" ", "-").lower()

    @property
    def skill_count(self) -> int:
        """Number of skills in this plugin."""
        return len(self.skills)

    @property
    def agent_count(self) -> int:
        """Number of agents in this plugin."""
        return len(self.agents)

    @property
    def connector_count(self) -> int:
        """Number of connectors in this plugin."""
        return len(self.connectors)

File: claude-code-agents-typescript/src/models/plugin-spec.ts
// src/models/plugin-spec.ts — Plugin specification models
import type { PluginName } from '../types.js';
import { toPluginName } from '../types.js';

// ── Skill Spec ──────────────────────────────────────────────────
export interface SkillSpec {
  readonly name: string;
  readonly description: string;
  readonly frontmatter: Readonly<Record<string, unknown>>;
  readonly content: string;
  readonly references: readonly string[];
  readonly scripts: readonly string[];
}

export function createSkillSpec(input: {
  readonly name: string;
  readonly description?: string;
  readonly frontmatter?: Record<string, unknown>;
  readonly content?: string;
  readonly references?: string[];
  readonly scripts?: string[];
}): SkillSpec {
  return {
    name: input.name,
    description: input.description ?? '',
    frontmatter: input.frontmatter ?? {},
    content: input.content ?? '',
    references: input.references ?? [],
    scripts: input.scripts ?? [],
  };
}

export function skillFileName(spec: SkillSpec): string {
  return `${spec.name}.md`;
}

// ── Agent Spec ──────────────────────────────────────────────────
export interface AgentSpec {
  readonly name: string;
  readonly description: string;
  readonly tools: readonly string[];
  readonly model: string;
  readonly systemPrompt: string;
}

export function createAgentSpec(input: {
  readonly name: string;
  readonly description?: string;
  readonly tools?: string[];
  readonly model?: string;
  readonly systemPrompt?: string;
}): AgentSpec {
  return {
    name: input.name,
    description: input.description ?? '',
    tools: input.tools ?? [],
    model: input.model ?? 'claude-sonnet-4-20250514',
    systemPrompt: input.systemPrompt ?? '',
  };
}

export function agentFileName(spec: AgentSpec): string {
  return `${spec.name}.md`;
}

// ── Connector Spec ──────────────────────────────────────────────
export type TransportType = 'stdio' | 'sse' | 'streamable-http';

export interface ConnectorSpec {
  readonly name: string;
  readonly type: TransportType;
  readonly serverConfig: Readonly<Record<string, unknown>>;
  readonly placeholderCategory: string;
}

export function createConnectorSpec(input: {
  readonly name: string;
  readonly type?: TransportType;
  readonly serverConfig?: Record<string, unknown>;
  readonly placeholderCategory?: string;
}): ConnectorSpec {
  return {
    name: input.name,
    type: input.type ?? 'stdio',
    serverConfig: input.serverConfig ?? {},
    placeholderCategory: input.placeholderCategory ?? '',
  };
}

export function hasPlaceholders(spec: ConnectorSpec): boolean {
  return JSON.stringify(spec.serverConfig).includes('~~');
}

// ── Plugin Spec ─────────────────────────────────────────────────
export interface PluginSpec {
  readonly name: PluginName;
  readonly version: string;
  readonly description: string;
  readonly author: string;
  readonly skills: readonly SkillSpec[];
  readonly agents: readonly AgentSpec[];
  readonly connectors: readonly ConnectorSpec[];
  readonly hooks: Readonly<Record<string, readonly Record<string, unknown>[]>>;
  readonly lspServers: readonly string[];
}

export function createPluginSpec(input: {
  readonly name: string;
  readonly version?: string;
  readonly description?: string;
  readonly author?: string;
  readonly skills?: SkillSpec[];
  readonly agents?: AgentSpec[];
  readonly connectors?: ConnectorSpec[];
  readonly hooks?: Record<string, Record<string, unknown>[]>;
  readonly lspServers?: string[];
}): PluginSpec {
  return {
    name: toPluginName(input.name),
    version: input.version ?? '0.1.0',
    description: input.description ?? '',
    author: input.author ?? '',
    skills: input.skills ?? [],
    agents: input.agents ?? [],
    connectors: input.connectors ?? [],
    hooks: input.hooks ?? {},
    lspServers: input.lspServers ?? [],
  };
}

export function pluginDirName(spec: PluginSpec): string {
  return (spec.name as string).replace(/ /g, '-').toLowerCase();
}

export function skillCount(spec: PluginSpec): number {
  return spec.skills.length;
}

export function agentCount(spec: PluginSpec): number {
  return spec.agents.length;
}

export function connectorCount(spec: PluginSpec): number {
  return spec.connectors.length;
}
